File: go/internal/fixtures/fixtures.go
// Package fixtures contains the shared test data used across both services.
//
// The tuples model this access scenario:
//
//	Alice → member of platformTeam
//	platformTeam#member → editor of productWorkspace
//	Bob → viewer of productWorkspace
//	roadmapDocument → lives in productWorkspace
//
// From these four tuples, the graph evaluator can derive:
//
//	Alice can_edit roadmapDocument  (via team → workspace editor → document)
//	Bob can_read roadmapDocument    (via workspace viewer → document viewer → can_read)
//	Casey cannot access roadmapDocument (no path in the graph)
package fixtures

import (
	"rebac-primer/internal/documents/adapters/authn"
	"rebac-primer/internal/shared"
)

// Named objects — use these in tests instead of raw strings.
var (
	Alice = shared.User("alice")
	Bob   = shared.User("bob")
	Casey = shared.User("casey")

	PlatformTeam     = shared.Team("platformTeam")
	ProductWorkspace = shared.Workspace("productWorkspace")
	RoadmapDocument  = shared.Document("roadmapDocument")
)

// DemoTokens maps demo bearer tokens to their claims.
// Mirrors typescript/src/demo/fixtures.ts demoTokens.
func DemoTokens() map[string]authn.TokenClaims {
	return map[string]authn.TokenClaims{
		"demo-token-alice": {Sub: "alice", Scopes: []string{"documents:read", "documents:write"}},
		"demo-token-bob":   {Sub: "bob", Scopes: []string{"documents:read"}},
		"demo-token-casey": {Sub: "casey", Scopes: []string{"documents:read"}},
	}
}

// SeedRelationshipTuples returns the four base tuples for the demo scenario.
func SeedRelationshipTuples() []shared.TupleKey {
	return []shared.TupleKey{
		// Alice is a member of platformTeam
		shared.Tuple(PlatformTeam, shared.RelationTeamMember, shared.Subject(Alice)),
		// platformTeam#member are editors of productWorkspace
		shared.Tuple(ProductWorkspace, shared.RelationWorkspaceEditor, shared.SubjectSet(PlatformTeam, shared.RelationTeamMember)),
		// Bob is a viewer of productWorkspace
		shared.Tuple(ProductWorkspace, shared.RelationWorkspaceViewer, shared.Subject(Bob)),
		// roadmapDocument lives in productWorkspace
		shared.Tuple(RoadmapDocument, shared.RelationDocumentWorkspace, shared.Subject(ProductWorkspace)),
	}
}

File: typescript/src/demo/fixtures.ts
// Demo seed data — the small cast of actors, objects, tokens, and policy tuples
// the example scenario is built around.
//
// This lives under src/ (not test/) because runtime entrypoints seed from it:
//   src/authz-service/index.ts      → seedPolicyTuples()
//   src/documents-service/index.ts  → demoTokens
//
// Tests also import it (re-exported via test/fixtures.ts), so the demo and the
// tests stay in sync. This mirrors the Go implementation's internal/fixtures
// package, which cmd/server/main.go and the Go tests both depend on.

import { subjectSet, team, tuple, user, workspace } from "../shared/rebac.ts";
import type { TupleKey } from "../shared/rebac.ts";

// ── Demo actors ───────────────────────────────────────────────────────────────

export const alice = user("alice");
export const bob   = user("bob");
export const casey = user("casey");

// ── Demo objects ──────────────────────────────────────────────────────────────

export const platformTeam     = team("platformTeam");
export const productWorkspace = workspace("productWorkspace");

// ── Demo bearer tokens ────────────────────────────────────────────────────────

// `satisfies` validates each entry against the claims shape while preserving the
// exact literal key type, so `keyof typeof demoTokens` is the union of real token
// strings rather than a plain `string`.
export const demoTokens = {
    "demo-token-alice": { sub: "alice", scopes: ["documents:read", "documents:write"] },
    "demo-token-bob":   { sub: "bob",   scopes: ["documents:read"] },
    "demo-token-casey": { sub: "casey", scopes: ["documents:read"] },
} satisfies Record<string, { sub: string; scopes: string[] }>;

// ── Policy tuples (workspace/team memberships) ────────────────────────────────
//
// These represent what the platform team writes to the authz service.
// Alice is a platform team member → editors of productWorkspace.
// Bob is a direct viewer of productWorkspace.

export const seedPolicyTuples = (): TupleKey[] => [
    tuple(platformTeam, "member", alice),
    tuple(productWorkspace, "editor", subjectSet(platformTeam, "member")),
    tuple(productWorkspace, "viewer", bob),
];
